### PR TITLE
Add hls tests

### PR DESCRIPTION
**Pull request by sulyi**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

In order to safeguard `HlsFD` new test added w/ various test cases covering the mentioned downloader's functionality.

- [x] enc
- [x] enc w/ iv
- [x] byte range
- [ ] ~byte range and enc~
- [x] byte range and enc w/ iv
- [x] webvtt
- [ ] media sequence
- [ ] ad segments (discontinuity, anvato, uplynk)

### CODE REVIEW

**Comment by sulyi** [comment]
Fixing commit messages (amend them w/ branch name).

**Comment by sulyi** [comment]
I remember having ideas for some more, but I forgot what they were. I may have left todos, I'll check, only rebased it. Debating to close it due to lack of interest.

**Comment by pukkandan** [comment]
wdym lack of interest? You just un-drafted it rn!

**Comment by sulyi** [comment]
> wdym lack of interest? You just un-drafted it rn!

Mainly on my end, I'm trying to resume work on it. See how it goes.

File: test/test_hls.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        try:
+            handle = subprocess.Popen(['ffmpeg', '-n', '-f', 'lavfi', '-re', '-i', 'testsrc=duration=0.65',
+                                       '-hls_init_time', '0.1s', '-hls_flags', 'split_by_time+single_file',
+                                       self.playlist],
+                                      cwd=DATA_DIR)
+        except OSError:
+            was_error = True
+        if was_error or handle.wait() != 0:
+            self.fail("Error occurred during generating files.")
+
```
[the remark this answers is not in the document]
Comment (sulyi) [reply test/test_hls.py:80]: Barely remember anything about this. I probably tried that and got copious amount of data, needless to be committed into repo.

Looking at todo it seems I've got stuck generating certain types of m3u8 files.

**Pull request closed without merging**